types/jest/index.d.ts

```diff
@@ -868,39 +841,23 @@ declare namespace jest {
         toHaveBeenCalledTimes(expected: number): R;
         /**
          * Ensure that a mock function is called with specific arguments.
-         *
-         * Optionally, you can provide a type for the expected arguments via a generic.
-         * Note that the type must be either an array or a tuple.
          */
-        // tslint:disable-next-line: no-unnecessary-generics
-        toHaveBeenCalledWith<E extends any[]>(...params: E): R;
+        toHaveBeenCalledWith(...args: T extends (...args: any) => any ? Parameters<T> : any[]): R;
```

Comment:
In a similar manner to toEqual and this will apply to other mock function matchers.

This change has introduced a typescript error for the test below that passes.
`
const fn = jest.fn((o:{a:number|undefined,b:number})=>{});
fn({a: undefined, b: 2})
expect( fn).toHaveBeenCalledWith({b: 2});
`

Reply:
Same as above, the test might pass but is semantically meaningless